File: backend/app/utils/response_formatter.py
import json
import re

def format_question_response(raw_response: str, default_id: str = "q-1", expected_difficulty: str = "Medium") -> dict:
    """
    Sanitize and parse LLM generated JSON response
    """
    try:
        clean_text = raw_response.strip()
        # Remove markdown fences like ```json ... ```
        if clean_text.startswith("```"):
            clean_text = re.sub(r"^```(?:json)?\s*", "", clean_text, flags=re.IGNORECASE)
            clean_text = re.sub(r"\s*```$", "", clean_text)

        parsed = json.loads(clean_text)

        options = parsed.get("options", [])
        if not isinstance(options, list) or len(options) != 4:
            options = ["Option A", "Option B", "Option C", "Option D"]

        return {
            "id": str(parsed.get("id", default_id)),
            "difficulty": str(parsed.get("difficulty", expected_difficulty)),
            "question": str(parsed.get("question", f"Sample {expected_difficulty} question regarding the topic")),
            "options": [str(opt) for opt in options],
            "correctAnswerIndex": int(parsed.get("correctAnswerIndex", 0)),
            "explanation": str(parsed.get("explanation", "Educational solution explanation."))
        }
    except Exception as err:
        print(f"[response_formatter] JSON parsing failed ({err}). Returning clean fallback structure.")
        return {
            "id": default_id,
            "difficulty": expected_difficulty,
            "question": f"Sample {expected_difficulty} question regarding the topic",
            "options": [
                "Correct answer demonstrating key concept",
                "Incorrect plausible distractor A",
                "Incorrect plausible distractor B",
                "Incorrect plausible distractor C"
            ],
            "correctAnswerIndex": 0,
            "explanation": f"This is a formatted question for {expected_difficulty} level."
        }

File: backend/src/utils/responseFormatter.js
/**
 * Sanitize and parse LLM generated response into structured Question JSON object
 */
export function formatQuestionResponse(rawResponse, defaultId = 'q-1', expectedDifficulty = 'Medium') {
  try {
    let cleanText = rawResponse.trim();
    
    // Remove markdown code block fences if present (e.g. ```json ... ```)
    if (cleanText.startsWith('```')) {
      cleanText = cleanText.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/, '');
    }

    const parsed = JSON.parse(cleanText);

    return {
      id: parsed.id || defaultId,
      difficulty: parsed.difficulty || expectedDifficulty,
      question: parsed.question || 'What is the correct answer for this topic question?',
      options: Array.isArray(parsed.options) && parsed.options.length === 4 
        ? parsed.options 
        : ['Option A', 'Option B', 'Option C', 'Option D'],
      correctAnswerIndex: typeof parsed.correctAnswerIndex === 'number' ? parsed.correctAnswerIndex : 0,
      explanation: parsed.explanation || 'No explanation provided.'
    };
  } catch (err) {
    console.error('Failed to parse LLM JSON response:', err, rawResponse);
    return {
      id: defaultId,
      difficulty: expectedDifficulty,
      question: `Sample ${expectedDifficulty} question regarding the chosen topic`,
      options: [
        'Correct answer demonstrating key concept',
        'Incorrect plausible distractor A',
        'Incorrect plausible distractor B',
        'Incorrect plausible distractor C'
      ],
      correctAnswerIndex: 0,
      explanation: `This is a fallback formatted question for ${expectedDifficulty} level.`
    };
  }
}
